tests/decorators/test_decorators.py

```diff
@@ -0,0 +1,802 @@
+#  Copyright (c) Microsoft Corporation. All rights reserved.
+#  Licensed under the MIT License.
+import unittest
+
+from azure.functions.decorators.core import DataType, AuthLevel, \
+    BindingDirection, AccessRights, Cardinality
+from azure.functions.decorators.function_app import FunctionApp
+from azure.functions.decorators._http import HttpTrigger, HttpMethod
+from azure.functions.decorators.timer import TimerTrigger
+from tests.decorators.testutils import assert_json, get_func
+
+
+class TestFunctionsApp(unittest.TestCase):
+    def setUp(self):
+        self.func_app = FunctionApp()
+
+    def test_route_is_function_name(self):
+        app = self.func_app
+        test_func_name = "dummy_function"
+
+        @app.function_name(test_func_name)
+        @app.route()
+        def dummy_func():
+            pass
+
+        func = get_func(self, app)
+
+        self.assertEqual(func.get_function_name(), test_func_name)
+        self.assertTrue(isinstance(func.get_trigger(), HttpTrigger))
+        self.assertTrue(func.get_trigger().route, test_func_name)
+
+    def test_route_is_python_function_name(self):
+        app = self.func_app
+
+        @app.route()
+        def dummy_func():
+            pass
+
+        func = get_func(self, app)
+
+        self.assertEqual(func.get_function_name(), "dummy_func")
+        self.assertTrue(isinstance(func.get_trigger(), HttpTrigger))
+        self.assertTrue(func.get_trigger().route, "dummy_func")
+
+    def test_route_is_custom(self):
+        app = self.func_app
+
+        @app.function_name("dummy_function")
+        @app.route("dummy")
+        def dummy_func():
+            pass
+
+        func = get_func(self, app)
+
+        self.assertEqual(func.get_function_name(), "dummy_function")
+        self.assertTrue(isinstance(func.get_trigger(), HttpTrigger))
+        self.assertTrue(func.get_trigger().route, "dummy")
+
+    def test_timer_trigger_default_args(self):
```

Comment:
Something to think about - as this test file is becoming bigger - have you explored some thoughts to breakup the tests into its own classes

A top-level `class TestDecorators` can have the basic stuff that individual test classes can inherit (eg the `self.func_app = FunctionApp()` as common setup)

and then `class TestHTTPDecorators(TestDecorators)`, `class TestTimerDecorators(TestDecorators)`, etc. can come in to make the handling of this long file easier. And also helps in test management on our end.

Reply:
Not for this PR though. CodeCoverage is more important currently.